zookeeper-server/src/main/java/org/apache/zookeeper/server/RequestThrottler.java

```diff
@@ -0,0 +1,273 @@
+     *
+     * When maxRequests = 0, throttling is disabled.
+     */
+    private static volatile int maxRequests =
+        Integer.getInteger("zookeeper.request_throttle_max_requests", 0);
+
+    /**
+     * The time (in milliseconds) this is the maximum time for which throttler
+     * thread may wait to be notified that it may proceed processing a request.
+     */
+    private static volatile int stallTime =
+        Integer.getInteger("zookeeper.request_throttle_stall_time", 100);
+
+    /**
+     * When true, the throttler will drop stale requests rather than issue
+     * them to the request pipeline. A stale request is a request sent by
+     * a connection that is now closed, and/or a request that will have a
+     * request latency higher than the sessionTimeout. The staleness of
+     * a request is tunable property, @see Request for details.
+     */
+    private static volatile boolean dropStaleRequests =
+        ZooKeeperServer.getBooleanProp("zookeeper.request_throttle_drop_stale", true);
+
+    public RequestThrottler(ZooKeeperServer zks) {
+        super("RequestThrottler", zks.getZooKeeperServerListener());
+        this.zks = zks;
+        this.stopping = false;
+        this.killed = false;
+    }
+
+    public static int getMaxRequests() {
+        return maxRequests;
+    }
+
+    public static void setMaxRequests(int requests) {
+        maxRequests = requests;
+    }
+
+    public static int getStallTime() {
+        return stallTime;
+    }
+
+    public static void setStallTime(int time) {
+        stallTime = time;
+    }
+
+    public static boolean getDropStaleRequests() {
+        return dropStaleRequests;
+    }
+
+    public static void setDropStaleRequests(boolean drop) {
+        dropStaleRequests = drop;
+    }
+
+    public static boolean enabled() {
+        return enabled;
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (true) {
+                Request request = submittedRequests.take();
+                if (Request.requestOfDeath == request) {
+                    break;
+                }
+
+                if (request.mustDrop()) {
+                    continue;
+                }
+
+                // Throttling is disabled when maxRequests = 0
+                if (maxRequests > 0) {
+                    while (!killed) {
+                        if (dropStaleRequests && request.isStale()) {
+                            // Note: this will close the connection
+                            dropRequest(request);
+                            ServerMetrics.getMetrics().STALE_REQUESTS_DROPPED.add(1);
+                            request = null;
+                            break;
+                        }
+                        if (zks.getInProcess() < maxRequests) {
+                            break;
+                        }
+                        throttleSleep(stallTime);
+                    }
+                }
+
+                if (killed) {
```

Comment:
this could be elevated to the top of the first while block, so we can break early and start draining the queue sooner. consider one case - we have a big queue of submitted requests and most of them must be dropped. If we leave the check here, we'll not break the loop until processing the last request (the request of death.).

Reply:
The check for killed is also in the inner while loop for each request in the queue. 
So it won't delay the shutdown.

Reply:
Assume while we are shutting down (`kill` is set to true), we have a large queue of requests, where each request is stale and must be dropped. With current code, the outer most `while (true) {` loop will try loop through every request in the queue, because L145 `if (request.mustDrop()) {` is evaluated as true for every request (which is stale). In this case, the check for `killed` in the inner while loop will not get executed. This was the reason I was suggesting to move the `if (killed) {` check before the `if (request.mustDrop()) {`.

Does this example make sense? Albeit this example somewhat hypothetical, having `if (killed) {` check earlier sounds no harm either.

Reply:
Yes, get it! Thank you for your detailed explanation! Fix on the way...

Reply:
@jhuan31  agree, add instead of moving is better.